File: 01/solutions.jl
#!/usr/bin/env julia

function calculateFull(mass::Integer) ::Integer
   return trunc(mass / 3) - 2
end


function otherFuel(mass::Integer) ::Integer
   submass = calculateFull(mass)
   if submass <= 0
      return 0
   end
   return submass + otherFuel(submass)
end


open("input", "r") do masses
   fuel = map(mass -> calculateFull(parse(Int, mass)), eachline(masses))
   print(fuel, "\n")
   print(sum(fuel), "\n")
end


open("input", "r") do masses
   fuel = map(mass -> otherFuel(parse(Int, mass)), masses)
   print(fuel, "\n")
   print(sum(fuel), "\n")
end

File: 01/solutions.py
#!/usr/bin/env python3


def calculateFull(mass):
    """
    Calculate the required fuel for a given mass.
    """
    return mass // 3 -2



def otherFuel(mass):
    """
    Recursively require the fuel require for a mass and its fuel.
    """
    submass = calculateFull(mass)
    if submass <= 0:
        return 0
    return submass + otherFuel(submass)




if __name__ == '__main__':
    with open('input', 'r') as f:
        fuel = sum(map(lambda x: calculateFull(int(x.strip())), f.readlines()))

    print('part1', fuel)

    with open('input', 'r') as f:
        fuel = sum(map(lambda x: otherFuel(int(x.strip())), f.readlines()))

    print('part2', fuel)

